File: javascript/DataStructures/CircularLinkedList.js
class Node {
    constructor(value) {
        this.data = value;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.tail = null;
    }

    pushFront(value) {
        let newNode = new Node(value);
        if (this.tail === null) {
            newNode.next = newNode;
            this.tail = newNode;
        } else {
            newNode.next = this.tail.next;
            this.tail.next = newNode;
        }
    }

    pushBack(value) {
        let newNode = new Node(value);
        if (this.tail === null) {
            newNode.next = newNode;
            this.tail = newNode;
        } else {
            newNode.next = this.tail.next;
            this.tail.next = newNode;
            this.tail = newNode;
        }
    }

    popFront() {
        if (this.tail === null) {
            throw new Error("Empty list. Cannot perform popFront!")
        }
        if (this.tail.next === this.tail) {
            this.tail = null;
        } else {
            this.tail.next = this.tail.next.next;
        }
    }

    popBack() {
        if (this.tail === null) {
            throw new Error("Empty list. Cannot perform popBack!")
        }
        if (this.tail.next === this.tail) {
            this.tail = null;
        } else {
            let curr = this.tail;
            while (curr.next !== this.tail) {
                curr = curr.next;
            }
            curr.next = curr.next.next;
            this.tail = curr;
        }
    }

    display() {
        if (this.tail === null) {
            throw new Error("Empty list. Cannot perform display!")
        }

        let curr = this.tail;
        let displayList = "";
        let lastNode = `${curr.data}->`;
        curr = curr.next;

        while (curr !== this.tail) {
            displayList += `${curr.data}->`;
            curr = curr.next;
        }
        displayList += lastNode;
        displayList += "this.head";
        console.log(displayList);
    }
}

// MAIN AREA //
const list = new LinkedList();
list.pushBack(1);
list.pushBack(3);
list.pushBack(4);
list.pushBack(5);
list.pushFront(0);

try {
    list.popFront();
} catch (err) {
    console.log(err);
}

try {
    list.popBack();
} catch (err) {
    console.log(err);
}

list.display();

File: go/DataStructures/CircularLinkedList/CircularLinkedList.go
package CircularLinkedList

import (
	"errors"
	"fmt"
)

type Node struct {
	data int
	next *Node
}

func NewNode(value int) *Node {
	return &Node{
		data: value,
		next: nil,
	}
}

type LinkedList struct {
	tail *Node
}

func NewLinkedList() *LinkedList {
	return &LinkedList{
		tail: nil,
	}
}

func (list *LinkedList) PushFront(value int) {
	newNode := NewNode(value)

	if list.IsEmpty() {
		newNode.next = newNode
		list.tail = newNode
	} else {
		newNode.next = list.tail.next
		list.tail.next = newNode
	}
}

func (list *LinkedList) PushBack(value int) {
	newNode := NewNode(value)

	if list.IsEmpty() {
		newNode.next = newNode
		list.tail = newNode
	} else {
		newNode.next = list.tail.next
		list.tail.next = newNode
		list.tail = newNode
	}
}

func (list *LinkedList) PopFront() error {
	if list.IsEmpty() {
		return errors.New("Empty list. Failed to perform PopFront")
	}
	if list.tail.next == list.tail {
		list.tail = nil
	} else {
		list.tail.next = list.tail.next.next
	}

	return nil
}

func (list *LinkedList) PopBack() error {
	if list.IsEmpty() {
		return errors.New("Empty list. Failed to perform PopBack")
	}
	if list.tail.next == list.tail {
		list.tail = nil
	} else {
		curr := list.tail.next

		for curr.next != list.tail {
			curr = curr.next
		}
		curr.next = curr.next.next
		list.tail = curr
	}

	return nil
}

func (list *LinkedList) IsEmpty() bool {
	return list.tail == nil
}

func (list *LinkedList) Display() {
	if !list.IsEmpty() {
		curr := list.tail.next
		lastNode := list.tail

		for curr != list.tail {
			fmt.Printf("%d->", curr.data)
			curr = curr.next
		}

		fmt.Printf("%d->", lastNode.data)
	}
	fmt.Println("NULL")
}
